### PR TITLE
Changed Hessian computations to analytical from numerical

### PR DESCRIPTION
**Pull request by maieneuro**
Testing to ensure correct values are being computed is still pending: especially for fe_poly_tensor.

### CODE REVIEW

**Comment by tamiko** [comment]
@maineuro I've scheduled a testsuite run

File: source/fe/fe_poly.cc [hunk 1]
```diff
@@ -56,11 +56,11 @@ FE_Poly<TensorProductPolynomials<1>,1,2>::fill_fe_values
...
         mapping.transform(fe_data.shape_gradients[k], data.shape_gradients[k],
                           mapping_data, mapping_covariant);
-    }
 
-  if (flags & update_hessians && cell_similarity != CellSimilarity::translation)
-    this->compute_2nd (mapping, cell, QProjector<1>::DataSetDescriptor::cell(),
-                       mapping_data, fe_data, data);
+      if (flags & update_hessians && cell_similarity != CellSimilarity::translation)
+        mapping.transform(fe_data.shape_hessians[k], data.shape_hessians[k],
+                          mapping_data, mapping_covariant_gradient);
```
Comment (bangerth) [inline source/fe/fe_poly.cc]: This is really all that is necessary? That's quite remarkable!

File: source/fe/fe_poly_tensor.cc [hunk 2]
```diff
@@ -426,162 +441,68 @@ FE_PolyTensor<POLY,dim,spacedim>::fill_fe_values (
   else if (mapping_type == mapping_nedelec)
     get_face_sign_change_nedelec (cell, this->dofs_per_face, sign_change);
 
+  // we only need to update element data if the current cell is not a
+  // translation of the previous one; or, even if it is a translation, if we
```
Comment (bangerth) [inline source/fe/fe_poly_tensor.cc]: There are a lot of changes in this function, to the point where it's difficult to read the patch. Do I see this right that you're just switching the inner loops and the outer switch statement?

File: source/fe/fe_poly.cc (hunk 1 above)
Comment (kronbichler) [inline source/fe/fe_poly.cc]: No, this can't be all that is needed in the general case (it is correct in the Cartesian/affine case, though). There should also be a term of the form `inv_jac_grad * fe_data.shape_gradients[k]`. I haven't checked all yet, though.

File: source/fe/fe_poly_tensor.cc [hunk 3]
```diff
@@ -1013,10 +1343,173 @@ FE_PolyTensor<POLY,dim,spacedim>::fill_fe_subface_values (
               Assert(false, ExcNotImplemented());
             }
         }
+      // update hessians
+      if (flags & update_hessians )
+
+        {
+          std::vector<Tensor<3, spacedim > > shape_grad_grads1 (n_q_points);
+          std::vector<Tensor<1, dim > > shape_hessians_for_fixed_indices (n_q_points);
+          std::vector<Tensor<1, dim > > transformed_shape_hessians_for_fixed_indices (n_q_points);
```
Comment (bangerth) [inline source/fe/fe_poly_tensor.cc]: These are just scratch arrays, and you're allocating dynamic memory here every time this function is called. Can you put them as `mutable` fields into the `InternalData` member of this class?

**Comment by bangerth** [comment]
I cannot express how unbelievably impressed I am at the speed with which you understood the code, and implemented the necessary changes. Incredible!

Before this is merged, there are two obstacles:
* I'd like to see how the tests come back. @tamiko -- can you post a link once they're through?
* I have a biggish patch almost ready that revises the argument list of all of the fill_fe_values and compute_2nd functions. This will clash with your patch. Not in a way that is going to be overly difficult to resolve, but it will nevertheless create conflicts. Would you allow me to put my patch in first, and then just look at the fallout? Do you know how to resolve merge conflicts?

(PS: My branch with the changes is here: https://github.com/bangerth/dealii/tree/further-FEValues-cleanups . I'll create a pull request out of it once I've ensured that the tests work.)

**Comment by maieneuro** [comment]
Please go ahead with your patch, I haven't worked with resolving merge conflicts before but I'm sure I'll manage. In the mean time I'll address the comments above.

**Comment by kronbichler** [comment]
@maieneuro You find the list of failing tests run by @tamiko at http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10590

It looks like parts of the second derivatives are indeed missing, I suspect those regarding the gradient of the inverse Jacobian of the transformation I mentioned above. (I see this because tests/matrix_free/get_functions_xxx provides a set of tests that compare up to second derivatives of the matrix-free evaluation with FEValues results; the test matrix_free/get_functions_cartesian.cc passes your patch, whereas the test matrix_free/get_functions_circle.cc and a few others based on non-constant Jacobians fail; those tests are actually supposed to test the correctness of some MatrixFree evaluations with FEValues as a reference, but you can also see them as a test of FEValues.)

**Comment by maieneuro** [comment]
I'm working on adding the missing term. What's the best way to access the Jacobian grads within fill_fe_values( )?

**Comment by bangerth** [comment]
I think it's one of the fields that is computed by the mapping, isn't it?

**Comment by bangerth** [comment]
(Or at least that *can* be computed by the mapping if they receive the corresponding update flag.)

**Comment by kronbichler** [comment]
Right now, we only have a flag for update_jacobian_gradients. I think we should introduce one for update_inverse_jacobian_gradients as well and put the data into the respective field (to be filled by Mapping). That data is then accessed in the code for the transformation.

**Comment by bangerth** [comment]
Good point. And specifying `update_inverse_jacobian_gradients` would then typically trigger `update_inverse_jacobian` and `update_jacobian_grad` and in fill_fe_values, the mappings would then simply do the appropriate product at the end to compute the gradient of the inverse from the inverse and the gradient?

**Comment by bangerth** [comment]
@maieneuro : Hamed, for reference, the gradient of shape functions (on the real cell) is given by
```
  J^{-1}(\hat x) \hat \nabla phi_i(\hat x)
```
where `\hat x` is the quadrature point in reference coordinates, and `\hat \nabla` are derivatives with regard to reference coordinates. Given this, the second derivatives are given by
```
  J^{-1}(\hat x) \hat \nabla [J^{-1}(\hat x) \hat \nabla phi_i(\hat x)]
```
which you can multiply out to yield
```
  J^{-1}(\hat x) J^{-1}(\hat x) [\hat \nabla^2 phi_i(\hat x)]
  +
  J^{-1}(\hat x) [\hat \nabla J^{-1}(\hat x)] [\hat \nabla phi_i(\hat x)]
```
(with appropriate contractions over tensor indices that you can get from the place Martin referenced). The first involves transforming the Hessians of the shape function on the reference cell, `\hat \nabla^2 phi_i(\hat x)` and I think that's the term that you implemented. The second term involves the derivative of the inverse Jacobian, `\hat \nabla J^{-1}(\hat x)` and I believe that this term was missing.

**Comment by maieneuro** [comment]
@bangerth   Yes, but this is equivalent to what I've implemented because
  
  $\hat \nabla J^{-1}(\hat x) = -J^{-1}(\hat x) \hat \nabla J(\hat x) J^{-1}(\hat x)$.
  
  The second term expands to 
  
  $J^{-1}(\hat x) [\hat \nabla J^{-1}(\hat x)] [\hat \nabla \phi_i(\hat x)] = J^{-1}(\hat x) \{ - [ J^{-1}(\hat x) \hat \nabla J(\hat x) ] [J^{-1}(\hat x)  \hat \nabla \phi_i(\hat x)]  \}$
  
 So, because as you pointed out $\nabla \phi_i(\hat x) = J^{-1}(\hat x) \hat \nabla \phi_i(\hat x)$,  the whole second derivative becomes a mapping_covariant_gradient transformation of 
  
  $[\hat \nabla^2 \phi_i(\hat x)] - \hat \nabla J(\hat x) \nabla \phi_i(\hat x)]$
  
  In other words, the gradient of the hessian inverse does not need to be computed directly. Am I missing something here?

**Comment by bangerth** [comment]
I didn't try to follow your formulas in detail (and you already posted a different patch anyway in #1272), but just to show that the formulas cannot be the same:
```
 J^{-1}(\hat x) J^{-1}(\hat x) [\hat \nabla^2 phi_i(\hat x)]
  +
  J^{-1}(\hat x) [\hat \nabla J^{-1}(\hat x)] [\hat \nabla phi_i(\hat x)]
```
This formula contains both the first and the second derivatives of the shape functions on the reference cell. No transformation can merge the second into the first term because it would imply that you can express the first derivatives in terms of the second, which is of course not possible.

**Comment by maieneuro** [comment]
Here's a derivation: 
https://drive.google.com/file/d/0B-7mnFLpdvqxbWluTGFjOHItdHc/view?usp=sharing

The reason the two terms are compatible is that the first derivative (which actually is in the real cell here) is multiplied by the jacobian gradients. The dimensions for the real coordinates cancel out and what remains has the dimensions of the second derivative of the shape function in the reference cell. 

I've changed the indexing and now the formula is giving sane results. Previously I was following the indexing given in the documentation of FEValuesBase, which suggests the index of jacobian_grads corresponding to the real cell is the last one. But it seems that it's actually the second index that corresponds to the real cell.

**Comment by bangerth** [comment]
Yes, I think this looks reasonable.
